internal/pkg/manifest/workload.go

```diff
@@ -441,14 +442,82 @@ func (p *PlacementArgs) isEmpty() bool {
 // PlacementString represents what types of subnets (public or private subnets) to place tasks.
 type PlacementString string
 
+// SecurityGroupsIDsOrConfig represents security groups attached to task. It supports unmarshalling
+// yaml which can either be of type SecurityGroupsConfig or a list of strings.
+type SecurityGroupsIDsOrConfig struct {
+	IDs            []string
+	AdvancedConfig SecurityGroupsConfig
+}
+
+func (s *SecurityGroupsIDsOrConfig) isEmpty() bool {
+	return len(s.IDs) == 0 && s.AdvancedConfig.isEmpty()
+}
+
+// SecurityGroupsConfig represents which security groups are attached to a task
+// and if default security group is applied.
+type SecurityGroupsConfig struct {
+	SecurityGroups []string `yaml:"groups"`
+	DenyDefault    *bool    `yaml:"deny_default"`
+}
+
+func (s *SecurityGroupsConfig) isEmpty() bool {
+	return len(s.SecurityGroups) == 0 && s.DenyDefault == nil
+}
+
+// UnmarshalYAML overrides the default YAML unmarshalling logic for the SecurityGroupsIDsOrConfig
+// struct, allowing it to perform more complex unmarshalling behavior.
+// This method implements the yaml.Unmarshaler (v3) interface.
+func (s *SecurityGroupsIDsOrConfig) UnmarshalYAML(value *yaml.Node) error {
+	if err := value.Decode(&s.AdvancedConfig); err != nil {
+		switch err.(type) {
+		case *yaml.TypeError:
+			break
+		default:
+			return err
+		}
+	}
+
+	if !s.AdvancedConfig.isEmpty() {
+		// Unmarshalled successfully to s.AdvancedConfig, unset s.IDs, and return.
+		s.IDs = nil
+		return nil
+	}
+
+	if err := value.Decode(&s.IDs); err != nil {
+		return errUnmarshalSecurityGroupOpts
+	}
+	return nil
+}
+
+// GetIDs returns security groups from SecurityGroupsIDsOrConfig that are attached to task.
+// nil is returned if no security groups are specified.
+func (s *SecurityGroupsIDsOrConfig) GetIDs() []string {
```

Comment:
@Lou1415926 suggestion sounds good but there's a conflict.
changing method to `IDs` conflicts with [field](https://github.com/gautam-nutalapati/copilot-cli/blob/83c3e587ac68c18ec7fab4ba0e6428774aafcf60/internal/pkg/manifest/workload.go#L448) in struct:
```
type SecurityGroupsIDsOrConfig struct {
	IDs            []string
	AdvancedConfig SecurityGroupsConfig
}
```
I can't change IDs to lowercase as its accessed in [test](https://github.com/gautam-nutalapati/copilot-cli/blob/83c3e587ac68c18ec7fab4ba0e6428774aafcf60/internal/pkg/deploy/cloudformation/stack/backend_svc_test.go#L386) of `stack` module.

I can make test refer to AdvancedConfig, and change `IDs` field to `ids`, and then rename the method as you suggested, but that looks hacky.
Or we can rename IDs for `identifiers`. 
Let me know what sounds better, or any other options that'll be good.

Reply:
> I can't change IDs to lowercase as its accessed in [test](https://github.com/gautam-nutalapati/copilot-cli/blob/83c3e587ac68c18ec7fab4ba0e6428774aafcf60/internal/pkg/deploy/cloudformation/stack/backend_svc_test.go#L386) of stack module.

Yeah not only that, but also that the `mergo` package used for environment overrides only works for exported field. If it was private it'd just skip over it without merging 😓.
 
I think we can keep the `GetIDs` naming then!!  It could also be argued that the naming conventions was for `Getter` that gets private fields, which is not particularly what we are doing here. Thank you for digging into it!